ui-manchette-with-spacetimechart/src/helpers.ts

```diff
@@ -154,17 +159,32 @@ export const getScales = (
   ];
 };
 
+export const zoomValueToTimeScale = (slider: number) =>
+  MIN_ZOOM_MS_PER_PX * Math.pow(MAX_ZOOM_MS_PER_PX / MIN_ZOOM_MS_PER_PX, slider / 100);
+
+export const timeScaleToZoomValue = (timeScale: number) =>
+  (100 * Math.log(timeScale / MIN_ZOOM_MS_PER_PX)) /
+  Math.log(MAX_ZOOM_MS_PER_PX / MIN_ZOOM_MS_PER_PX);
+
 /** Zoom on X axis and center on the mouse position */
 export const zoomX = (
   currentXZoom: number,
   currentXOffset: number,
-  { delta, position: { x } }: Parameters<NonNullable<SpaceTimeChartProps['onZoom']>>[0]
+  newXZoom: number,
+  position: number
 ) => {
-  const xZoom = Math.min(Math.max(currentXZoom * (1 + delta / 10), MIN_ZOOM_X), MAX_ZOOM_X);
+  const boundedXZoom = clamp(newXZoom, MIN_ZOOM_X, MAX_ZOOM_X);
+  const oldTimeScale = zoomValueToTimeScale(currentXZoom);
+  const newTimeScale = zoomValueToTimeScale(boundedXZoom);
+  // by default zooming expands the graph around timeOrigin
+  // changing the offset to timeOriginShift alone keeps the left border in place.
+  // subtracting centerShift keeps the center in place
+  // (xOffset = how many px the timeOrigin is shifted to the right)
+  const timeOriginShift = (currentXOffset * oldTimeScale) / newTimeScale;
+  const centerShift = (position * oldTimeScale - position * newTimeScale) / newTimeScale;
+  const newOffset = timeOriginShift - centerShift;
   return {
-    xZoom,
-    // Adjust zoom level relatively to the input delta value:
-    // These lines are here to center the zoom on the mouse position:
-    xOffset: x - ((x - currentXOffset) / currentXZoom) * xZoom,
```

Comment:
Was this old logic not working somehow? Naively, I'd think it'd work fine with the new zoom (modulo some conversions)…

Reply:
You’re right, I just needed a few adjustment to make it work (timeScale has an inverse relationship to zoomValue, previously we didn’t have inverse relationship)

much better !

`const newOffset = position - ((position - state.xOffset) * oldTimeScale) / newTimeScale;`